vermouth/tests/test_ff_files.py

```diff
@@ -0,0 +1,1019 @@
+        (
+            "BB SC1 1 dist(BB,SC1)",
+            ['1', vermouth.molecule.ParamDistance(['BB', 'SC1'])],
+        ),
+        (
+            "BB SC1 2 dist(BB,SC1|.3f) 6",
+            ['2', vermouth.molecule.ParamDistance(['BB', 'SC1'], format_spec='.3f'), '6'],
+        ),
+    ))
+    def test_param_effector(interaction_line, expected):
+        """
+        Parameter effectors are detected and instanciated.
+        """
+        lines = """
+        [ moleculetype ]
+        GLY 1
+
+        [ atoms ]
+        1 P4 1 ALA BB 1
+        2 P4 1 ALA SC1 1
+
+        [ bonds ]
+        {}
+        """
+        lines = textwrap.dedent(lines.format(interaction_line))
+        lines = lines.splitlines()
+
+        ff = vermouth.forcefield.ForceField(name='test_ff')
+        vermouth.ffinput.read_ff(lines, ff)
+        parameters = ff.blocks['GLY'].interactions['bonds'][0].parameters
+
+        assert parameters == expected
+
+    @staticmethod
+    def test_param_effector_fail():
+        """
+        A non existing parameter effector causes the parser to fail.
+        """
+        lines = """
+        [ moleculetype ]
+        GLY 1
+
+        [ atoms ]
+        1 P4 1 ALA BB 1
+        2 P4 1 ALA SC1 1
+
+        [ bonds ]
+        BB SC1 7 inexistant(BB,SC1)
+        """
+        lines = textwrap.dedent(lines)
+        lines = lines.splitlines()
+
+        ff = vermouth.forcefield.ForceField(name='test_ff')
+        with pytest.raises(IOError):
+            vermouth.ffinput.read_ff(lines, ff)
+
+
+class TestLink:
+    @staticmethod
+    def test_link():
+        """
+        The minimum link works.
+        """
+        lines = """
+        [ link ]
+
+        [link]
+        """
+        lines = textwrap.dedent(lines).splitlines()
+        ff = vermouth.forcefield.ForceField(name='test_ff')
+        vermouth.ffinput.read_ff(lines, ff)
+        assert len(ff.links) == 2
+
+    @staticmethod
+    @pytest.mark.parametrize('lines, atoms, interactions, edges', (
+        (  # Base case
+            """
+            [link]
+            [bonds]
+            BB SC1 1 0.3 100
+            """,
+            (
+                ('BB', {'atomname': 'BB', 'order': 0}),
+                ('SC1', {'atomname': 'SC1', 'order': 0}),
+            ),
+            {'bonds': [
+                vermouth.molecule.Interaction(
+                    atoms=['BB', 'SC1'],
+                    parameters=['1', '0.3', '100'],
+                    meta={},
+                ),
+            ]},
+            (('BB', 'SC1'),),
+        ),
+        (  # Base case with prefix
+            """
+            [link]
+            [bonds]
+            BB +SC1 1 0.3 100
+            """,
+            (
+                ('BB', {'atomname': 'BB', 'order': 0}),
+                ('+SC1', {'atomname': 'SC1', 'order': 1}),
+            ),
+            {'bonds': [
+                vermouth.molecule.Interaction(
+                    atoms=['BB', '+SC1'],
```

Comment:
Is it desirable it refers to atoms including the prefix?

Reply:
I am not fully convinced about it, but it is what we have been doing so far. It causes some weirdness is some edge cases (not errors, just possible traps when writing a link), but it assures that a key always refers to the same node and it makes some matching easier latter on.

Anyway, changing that is beyond the scope of the PR.

Reply:
I guess it depends on what the node key is. I agree that it's out of scope either way.

Reply:
And then it starts to make sense, since you can make a link `{BB, +BB}`. It's kind of convenient to have the node key BB or +BB, makes finding things easier, and it can't be the atomname, since that's not unique.

Reply:
I think that was the initial motivation, indeed. It bites you in rare cases where you define the order explicitly instead of by prefix:

```
[ link ]
[ atoms ]
SC1 {"order": ">", "custom": "for the demo"}
[ bonds ]
BB  >SC1
```

In the example above, one may think that we created the key `SC1` in the `[ atoms ]` section. Instead, we created the `>SC1` key. It is a niche corner case, but it exists. Shit, I'll have to document it :cry: